src/Engine/ProtoCore/Lang/CallSite.cs

```diff
@@ -1671,6 +1677,16 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
                 //We will call the subsequent reductions n times
                 int cartIndex = ri.CartesianIndex;
 
+                //We determine if the input parameters are homogeneous to set the final function endpoint once
+                if (cartIndex == 0)
+                {
+                    List<StackValue> finalFormalParameters = new List<StackValue>();
+                    var homogeneous = AreParametersHomogeneous(formalParameters, runtimeCore, finalFormalParameters);
```

Comment:
this has a worst case cost of 2*n - and I don't know much about the VM - so a question - by the time we are making a function call here - have we already marshaled this data?